### ISSUE TITLE
TYPE_CHECKING from typing module ignored

### ISSUE THREAD

**Issue by mcfreis** [Nuitka/Nuitka#1244]
* Problem or Bug:

    "imports" conditionally added in modules with TYPE_CHECKING are really imported which shouldn't be.

```
from __future__ import annotations
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from ....common.devices.base.device import DEVICE_TYPES  # this should only be imported by IDEs for type hinting, not at runtime

from ..devices.controller import Controller
from .signal import Signal

class Api(object):

    def __init__(self):
        self.app = None
        self.mainwindow = None
        self.controller: Optional[Controller] = None
        self.device: Optional[DEVICE_TYPES] = None
        self.newDevice: Optional[Signal] = None
```

* Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)

  wine python3 -m nuitka --version
  0.6.17.2
  Commercial: None
  Python: 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 10:41:24) [MSC v.1900 64 bit (AMD64)]
  Executable: C:\Program Files\Python37\python3.exe
  OS: Windows
  Arch: x86_64

* How did you install Nuitka and Python

  with pip3

**Comment by kayhayen** [comment]
This is now on factory branch, but the value tracing can easily not apply the optimization yet. I am working to make these kinds of module variables not become unknown when outside code is run, but it's not yet there, so this will only work relatively short after the import.

**Comment by kayhayen** [comment]
https://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
Yes, I think the import after the `TYPE_CHECKING` import currently makes all module variables considered "unknown", i.e. could be deleted, or any other value, such than the if statement then ignores it.

On my agenda is to excempt the so called hard imports and potentially other kinds of values from this level of paranoia, but it will take a bit of changes. Above code currently optimized into:

```TYPE_CHECKING = False
from .setiInterfaceParams import SetiInterfaceParams as InterfaceItem
if TYPE_CHECKING:
```

Obviously, that looses the information of origin, and therefore module variables need to be split into two kinds, one which is more trustworthy than the other, and then of course, to allow users to accept that all module variables of a given module are considered to not have outside writes.

So therefore this will have to wait until I figure out the refactorings necessary. btw, the same issue is affecting `sys.version_info` and `importlib.import_module`  and many other important things, so it's certainly going to come, but it's also not easy. If you rearrage above code though, moving the if statement up, it would work already now. Naturally not a solution that is practical.

**Comment by kayhayen** [comment]
There is now tracing enhancements, that should make sure this works on the module level. For uses inside functions, where that import is not done locally, it still won't work, but that will follow too.

**Comment by kayhayen** [comment]
This is part of the 0.6.18 release. On factory, the function level usage is implemented and will be in 0.6.19, use that if you encounter problems with this not being optimized yet.
